consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java

```diff
@@ -0,0 +1,180 @@
+
+    validators.add(Util.publicKeyToAddress(proposerKey.getPublicKey()));
+    validators.add(Util.publicKeyToAddress(validator1Key.getPublicKey()));
+    validators.add(Util.publicKeyToAddress(validator2Key.getPublicKey()));
+
+    final ProtocolContext<IbftContext> protocolContext =
+        new ProtocolContext<>(
+            mock(MutableBlockchain.class), mock(WorldStateArchive.class), mock(IbftContext.class));
+
+    @SuppressWarnings("unchecked")
+    BlockHeaderValidator<IbftContext> headerValidator =
+        (BlockHeaderValidator<IbftContext>) mock(BlockHeaderValidator.class);
+    BlockHeader parentHeader = mock(BlockHeader.class);
+
+    Map<ConsensusRoundIdentifier, MessageValidator> messageValidators = Maps.newHashMap();
+
+    messageValidators.put(
+        ri1,
+        new MessageValidator(
+            validators,
+            Util.publicKeyToAddress(proposerKey.getPublicKey()),
+            ri1,
+            headerValidator,
+            protocolContext,
+            parentHeader));
+
+    messageValidators.put(
+        ri2,
+        new MessageValidator(
+            validators,
+            Util.publicKeyToAddress(validator1Key.getPublicKey()),
+            ri2,
+            headerValidator,
+            protocolContext,
+            parentHeader));
+
+    messageValidators.put(
+        ri3,
+        new MessageValidator(
+            validators,
+            Util.publicKeyToAddress(validator2Key.getPublicKey()),
+            ri3,
+            headerValidator,
+            protocolContext,
+            parentHeader));
+
+    manager = new RoundChangeManager(2, validators, messageValidators::get);
+  }
+
+  private SignedData<RoundChangePayload> makeRoundChangeMessage(
+      final KeyPair key, final ConsensusRoundIdentifier round) {
+    MessageFactory messageFactory = new MessageFactory(key);
+    return messageFactory.createSignedRoundChangePayload(round, Optional.empty());
+  }
+
+  @Test
+  public void rejectsInvalidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(nonValidatorKey, ri1);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.roundChangeCache.get(ri1)).isNull();
+  }
+
+  @Test
+  public void acceptsValidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(proposerKey, ri2);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.roundChangeCache.get(ri2).receivedMessages.size()).isEqualTo(1);
+  }
+
+  @Test
+  public void doesntAcceptDuplicateValidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(proposerKey, ri2);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.roundChangeCache.get(ri2).receivedMessages.size()).isEqualTo(1);
+  }
+
```

Comment:
Do we need a similar test to check that we don't accept further round change message after reaching the threshold?

Reply:
If we stick with the actioned behaviour yes. Depending on the outcome to @rain-on's earlier comment I'll add a test or adjust.

Reply:
Test is added